tests/test_intl/test_intl.py

```diff
@@ -681,50 +680,188 @@ def test_translation_progress_classes_true(app):
     assert len(doctree[0]) == 20
 
 
+class _MockClock:
+    """Object for mocking :func:`time.time_ns` (if needed).
+
+    Use :meth:`sleep` to make this specific clock sleep for some time.
+    """
+
+    def time(self) -> int:
+        """Nanosecond since 'fake' epoch."""
+        raise NotImplementedError
+
+    def sleep(self, ds: float) -> None:
+        """Sleep *ds* seconds."""
+        raise NotImplementedError
+
+
+class _MockWindowsClock(_MockClock):
+    """Object for mocking :func:`time.time_ns` on Windows platforms.
+
+    The result is in 'nanoseconds' but with a microsecond resolution
+    so that the division by 1_000 does not cause rounding issues.
+    """
+
+    def __init__(self) -> None:
+        self.us: int = 0  # current microsecond 'tick'
+
+    def time(self) -> int:
+        ret = 1_000 * self.us
+        self.us += 1
+        return ret
+
+    def sleep(self, ds: float) -> None:
+        self.us += int(ds * 1e6)
+
+
+class _MockUnixClock(_MockClock):
+    """Object for mocking :func:`time.time_ns` on Unix platforms.
+
+    Since nothing is needed for Unix platforms, this object acts as
+    a proxy so that the API is the same as :class:`_MockWindowsClock`.
+    """
+
+    def time(self) -> int:
+        return time.time_ns()
+
+    def sleep(self, ds: float) -> None:
+        time.sleep(ds)
+
+
+@pytest.fixture()
+def mock_time_and_i18n(
+    monkeypatch: pytest.MonkeyPatch,
+) -> tuple[pytest.MonkeyPatch, _MockClock]:
+    from sphinx.util.i18n import CatalogInfo
+
+    # save the 'original' definition
+    catalog_write_mo = CatalogInfo.write_mo
+
+    def mock_write_mo(self, locale, use_fuzzy=False):
+        catalog_write_mo(self, locale, use_fuzzy)
+        # ensure that the .mo file being written has a correct fake timestamp
+        _set_mtime_ns(self.mo_path, time.time_ns())
```

Comment:
@picnixz I was thinking about this part of the PR again - this is where most of the repair happens, if I understand correctly.

It does two things:

  * Retrieves a mocked, monotonic timestamp by calling `time.time_ns` - this is guaranteed to be less than the next value retrieved.
  * Ensures that the timestamp is the one written to the modified-time on the `.mo` file on the filesystem.

Instead of patching over this in a test case where it occurs -- I wonder whether we should move the platform-specific-fix (even though it is ugly) into the Sphinx application code?

I tried that with a patch that includes:

```diff
diff --git a/sphinx/util/i18n.py b/sphinx/util/i18n.py
index 004b5d85f..dd0ff1cc5 100644
--- a/sphinx/util/i18n.py
+++ b/sphinx/util/i18n.py
@@ -4,6 +4,7 @@ from __future__ import annotations
 
--- a/sphinx/util/i18n.py
+++ b/sphinx/util/i18n.py
@@ -4,6 +4,7 @@ from __future__ import annotations
 
 import os
 import re
+import time
 from datetime import datetime, timezone
 from os import path
 from typing import TYPE_CHECKING, Callable, NamedTuple
@@ -68,6 +69,16 @@ class CatalogInfo(LocaleFileInfoBase):
                 write_mo(file_mo, po, use_fuzzy)
             except Exception as exc:
                 logger.warning(__('writing error: %s, %s'), self.mo_path, exc)
+                return
+
+        # Allow 318us of time to pass so that when the env.all_docs dependency
+        # graph is updated, Windows always records a timestamp for the catalog
+        # that is strictly greater-than the written .mo file's mtime_ns.
+        #
+        # Ref: https://github.com/sphinx-doc/sphinx/issues/11941
+        # Ref: https://peps.python.org/pep-0564/#windows
+        if os.name != 'posix':
+            time.sleep(3.18e-7)
 
 
 class CatalogRepository:
```

The intended benefit of doing that would be twofold: it would allow us to avoid the mocking/patching in the tests (simpler), and could (very rarely, I expect) prevent the same issue happening if for some reason projects are rapidly rebuilt on Windows hosts.

(I wouldn't have been able to figure out how to do that without experimenting with the patched test code - and I'm still not 100% certain that it'd be sensible, but it's an idea)

Reply:
I prefer not touching the code base at all. For release builds, the issue is that we detect documents as outdated while the are not necessarily really outdated. So it doesn't hurt to rebuild the docs. 

However, for tests, it's annoying that they fail since we run them quite often. 

Concerning the mock itself, every builder (even the dummy one) create those po/mo files. However, since they write something on the disk, the timestamps are given by the OS directly and the 'time' is given by the OS. However, I want the time always be given by this monotonic fake clock (what I essentially do is forcing 'time' informations that are not thread-relevant to be simulated by my fake clock).

We could have added a sleep after writing the po/mo files but I don't know how much files would be added. I think that there are MO files for *every* different RST file so for projects with thousands of files, this small sleep might be relevant. Note that `time.sleep` might come with a slight overhead and it's not guaranteed that you're actually sleeping what you are requesting (it may be less or more, it depends on the platform). Thus, to be on the safe side, we would probably have something like `time.sleep(0.001)` (and I don't think we parallelize this task).

For what it's worth, I think we shouldn't put delays in the release if they are only there to 'patch' tests. Also, note that we explicitly choose to have microsecond resolution and also allow for rounding errors (see `sphinx.environment._last_modified_time`).

Reply:
Ok, yep - I agree with that reasoning.

If in future someone _does_ happen to encounter this issue in practice on a Windows-based host, then ideally they can provide a bugreport and test coverage to demonstrate that, and _then_ it could be fixed in-application (as opposed to the other way round, like here, where we have an intermittent problem unrelated to the purpose of the test, and changing the app to cater for that does seem odd).